### ISSUE TITLE
[strict provenance] Provide a way to "create allocations at a fixed address"

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#98593]
This issue is part of the Strict Provenance Experiment - https://github.com/rust-lang/rust/issues/95228

On some platforms it makes sense to just take a hard-coded address, cast it to a pointer, and starting working with that pointer -- because there are external environment assumptions that say that certain things are provided at certain addresses.

This is perfectly fine for Rust as long as that memory is entirely *disjoint* from all the memory that Rust understands (`static` globals, stack allocations, heap allocations). Basically we can think of there being a single hard-coded provenance for "all the memory that is disjoint from the Abstract Machine", and that is the provenance we would like to use for these pointers created from hard-coded addresses. These restrictions make this operation a lot easier to specify than [`from_exposed_addr`](https://doc.rust-lang.org/nightly/std/ptr/fn.from_exposed_addr.html).

In the spirit of the Strict Provenance APIs, that means we probably want a function that does this, and that we can attach suitable documentation to. There are some open questions for the syntax and semantics of that function:
- What should it be called? `make_alloc`, `assume_alloc`, `hard_coded_alloc`? I am leaning towards something with "alloc" because this function is basically like an allocator, except that you tell it at which address to allocate and you have to promise that that is Okay To Do.
- It should definitely take a `usize` for the address and return `*mut T`. Should it also take a size, saying how large this assumed allocation is?
- Should each invocation return a fresh provenance, or should they all return the same provenance? Probably the latter; using a fresh provenance comes with a bunch of restrictions that I doubt such low-level code will be happy about.
- (other things I have not thought of)

Cc @Lokathor who keeps mentioning this usecase every time I want to ban int2ptr casts. ;)
Tagging WG-embedded since that's where this kind of stuff mostly happens (AFAIK)

**Comment by RalfJung** [comment]
Until such a function exists, one can use `from_exposed_addr_mut` for this purpose. The "provenance for all things outside the Abstract Machine" is definitely exposed and can hence be picked up by that function.

**Comment by Lokathor** [comment]
One thing I'd like to mention at the start is that hardware addresses should not be modeled as "owned" by anything inside Rust. They're generally volatile, and generally what you read from them might change without Rust doing anything on its end. At best we should think of them as Rust *sharing* the address with some external force.

Because of this, I think a name like "alloc" would be a misstep. People have the notion/understanding that when you alloc some memory you own it while it's live, and then you (or some drop glue) explicitly does a "free", and and it's dead. That's not really how hardware access works at all, so for a new situation we pick a new name.

Particularly, lots of existing and well working rust code doesn't bother with a "free" step at all when using hardware pointers. You just make up a pointer that you know is good, you read or write, and then you forget the pointer, and someone else might even be holding a pointer to that same location while you're doing all this, and it's all fine. If a person *wants* to build an ownership-style API on top of this using the borrow checker they can do that (there's several such examples), but that's separate from the base requirements of the abstract machine interacting with the hardware.

**Comment by clarfonthey** [comment]
Perhaps directly labelling these addresses as volatile might be the best way to go? This would also help for FFI as well, since you could make the distinction between an address for something that is "owned" inside the Rust code until passed back via FFI, or something that could concurrently be modified outside of Rust, by hardware or another thread.

**Comment by RalfJung** [comment]
@Lokathor what you say also confirms my thoughts above that we want to use the same provenance for all calls to this function, and not generate a fresh one. That provenance is assumed to be exposed from the start, so it is basically public and can change any time control is given to other code (including via another thread).

Using volatile accesses is a separate concern; Rust assumes for *all* pointers that if you do 2 non-atomic loads immediately after one another, they will give the same result -- there is no good way to opt-out of that assumption just for a specific provenance; this needs a more explicit marker at the relevant access. (I.e., making it volatile.)

> Perhaps directly labelling these addresses as volatile might be the best way to go?

There is no such thing as a volatile address. Being volatile is the property of an *access*.
If you want to ensure that all accesses to an address are volatile, that's fine, but this can't be automatic or else Rust would have to assume that *any* raw pointer access might have to be volatile.

There are [plenty of discussions](https://github.com/rust-lang/unsafe-code-guidelines/issues?q=is%3Aissue+is%3Aopen+volatile) around the semantics of volatile, their interaction with reference types, and so on; I would like to keep them out of here. This thread is solely about the name, signature, and provenance interaction of the function that is intended to replace those casts.

**Comment by Lokathor** [comment]
> so it is basically public and can change any time control is given to other code (including via another thread).

Nit: a hardware address can change _even without_ control going to other code.

It could be modeled that the hardware is another thread and that the hardware thread has started before the abstract machine does (similar to how static values are initialized before the abstract machine begins), if that suits you.

> There is no such thing as a volatile address. Being volatile is the property of an access.

But that doesn't matter at all? There's no such thing as a wrapping integer, wrapping is a property of an integer operation. And yet we have a "wrapping" type for our integers that makes all the normal operations for the type into the wrapping version. It wouldn't be a world breaker to just have a volatile pointer type (mostly for addresses outside the AM, but not exclusively) where `*ptr` is a `read_volatile` or `write_volatile` operation instead of a standard `read` or `write` operation. And then it would be pretty clear in a function signature "oh this pointer isn't the normal kind of pointer".

**Comment by RalfJung** [comment]
> Nit: a hardware address can change even without control going to other code.

I am well aware. That's what `volatile` is for.

And I think we can very easily not make this the 50th thread about volatile. Just focus on the provenance concern. They are entirely orthogonal. We do *not* have to co-design this with a volatile pointer type, and you are *not* helping this cause by ting all these issues in one big knot that is impossible to tackle incrementally.

**Comment by RalfJung** [comment]
There are plenty of cases where you want access at a particular address that has *nothing to do with volatile*. For example, you might write a kernel and know that the firmware put some data structure with important info at a hard-coded address.

So, this API certainly should not be tied to volatile in any way. That's why I marked the previous posts about volatile as off-topic.

**Comment by adamgreig** [comment]
I'll raise this at the next embedded WG meeting (tues 28th 20:00 CEST) to see if anyone else has thoughts.

> It should definitely take a usize for the address and return *mut T. Should it also take a size, saying how large this assumed allocation is?

I'd think so, we will often have a pointer to a large struct or array or quite possibly an entire memory region that is only for MMIO with fixed addresses (e.g. on Cortex-M, 0x4000_0000 through 0x5FFF_FFFF is all peripheral MMIO).

**Comment by RalfJung** [comment]
> I'd think so, we will often have a pointer to a large struct or array or quite possibly an entire memory region that is only for MMIO with fixed addresses (e.g. on Cortex-M, 0x4000_0000 through 0x5FFF_FFFF is all peripheral MMIO).

And then would you want it to be UB to use that pointer outside the given range? Or what would the exact consequence be of setting a particular size?

**Comment by adamgreig** [comment]
If it didn't take a size, would you assume a single word provenance or the whole memory, like an escaped pointer? I was imagining you'd want to be able to restrict it to the smallest usable provenance, which is generally known statically or easily bounded to a memory region.

If there's not really any advantage to the compiler to knowing the possible valid range of the pointer, then I suppose just taking the address and giving it a whole-memory provenance is a simpler API.

My guess for a typical embedded use-case is you'd create one of these pointers for each instance of a peripheral on your chip, so each has a non-overlapping size of say <100 words, and probably create them on-demand each time you accessed the peripheral.

**Comment by RalfJung** [comment]
> If it didn't take a size, would you assume a single word provenance or the whole memory, like an escaped pointer? 

It would be allowed to access all memory that is outside of what the Abstract Machine knows about (`static`, stack, heap).

> probably create them on-demand each time you accessed the peripheral.

This means multiple calls to that function can overlap anyway, so a size does not give any disjointness guarantees.

**Comment by Lokathor** [comment]
A size/span restriction does not seem useful.

**Comment by adamgreig** [comment]
> It would be allowed to access all memory that is outside of what the Abstract Machine knows about (static, stack, heap, anything internal to the implementation like compiler-generated data).

Right, I should have read the issue more carefully. I agree then, it doesn't seem like there'd be any benefit to giving a size, even if it is known.

For naming, I'm also not entirely sure about saying `alloc`; maybe the thing to express is that we're making a pointer that shares some global disjoint-from-known-universe provenance with all other similarly-created pointers? I think `extern` has too many other meanings, but if it didn't, something like `make_extern` or `assume_extern` could work.

**Comment by jamesmunns** [comment]
Thinking a bit about the use cases here, just to make sure I understand:

* Fixed memory map addresses (e.g. MMIO pointers)
* Runtime mapped regions (virtual memory mapping?), created by the code itself
* Runtime mapped regions done indirectly from the code (i.e. `mmap` - this might have a separate function for provenance? Or is this just handled by "pointer escaping" in the compiler currently?)

You mention specifically "for these pointers created from hard-coded addresses", would the latter two runtime-y cases also follow/benefit from this API?

If all three are relevant, I'd propose `make_mapped` or `assume_mapped` (or maybe `make|assume_extern_mapped`, as these all seem to derive from cases where we have memory that has been mapped (either statically or at runtime) into an address space outside of Rust's control.

**Comment by Lokathor** [comment]
"external address" seems like a good term, `<*mut T>::from_extern` seems maybe like a fine way to make them.

**Comment by RalfJung** [comment]
Yeah I considered "extern", too -- but the term is awfully overloaded.

> You mention specifically "for these pointers created from hard-coded addresses", would the latter two runtime-y cases also follow/benefit from this API?

This is specifically to replace current uses of `int as *mut PTR`. Functions like `mmap` already return a pointer and that one for all intents and purposes will have a fresh provenance -- they are like `malloc`, I think.

**Comment by jamesmunns** [comment]
That makes sense for the third point!

My second item is more for *implementing* an operating system, where you might want to map a page, zero it, then pass it on to "userspace".

If you were implementing ASLR for example, you'd (more or less) be generating the base mapped address from an RNG, which would probably have come from some kind of integer rather than pointer. That being said, the OS itself probably sees the physical memory address the same as case one. Once the OS establishes the pointer, and passes it to userspace (where userspace is in case three), you're right that userspace doesn't have to re-establish provenance, I guess.

**Comment by Dirbaio** [comment]
"foreign" maybe? so there's "program memory" and "foreign memory". The function could be `ptr::from_foreign_addr()`.

**Comment by RalfJung** [comment]
@jamesmunns from the kernel perspective this is memory outside of what Rust knows about, so if there are int2ptr casts there then using this function probably makes sense.

However now that if you e.g. implement a global allocator, then the memory return by that *is* considered to be "known to Rust" and thus must not be accessed with that "external"/"foreign" provenance.

**Comment by A1-Triard** [comment]
> Functions like `mmap` already return a pointer and that one for all intents and purposes will have a fresh provenance -- they are like `malloc`, I think.

But what if I want to perform `mmap` syscall using asm? Looks like it seems you need a way to create a new, unique provenance for a such case.

**Comment by Lokathor** [comment]
inline asm provenance creation rules *should* be the same as FFI provenance creation rules (whatever those are).

I don't know if that's fully decided yet, but in all other situations inline asm works "basically like FFI with a weird calling convention", and I see no reason to follow from that convention.

**Comment by RalfJung** [comment]
`asm!` block semantics are axiomatized anyway. You just have to be able to *describe* what happens in terms of the Abstract Machine state (in a way that is consistent with the actions of the `asm!` on the physical state); so that description can just state that it generates a fresh provenance since that is something the Abstract Machine can do.

**Comment by programmerjake** [comment]
one other case to consider is where the program picks an address, and then memmaps to that address, using something like [`MAP_FIXED`](https://manpages.debian.org/bullseye/manpages-dev/mmap.2.en.html#MAP_FIXED):
```rust
let addr: *mut u8 = 0x12340000 as *mut u8; // pick an arbitrary page-aligned address
let addr = mmap(addr as *mut c_void, 0x10000, PROT_READ | PROT_WRITE, MAP_FIXED | MAP_PRIVATE, fd, 0);
```

imho `mmap` here should return the same provenance as the `addr` argument even though that `addr` argument wasn't valid until the `mmap` call.

**Comment by RalfJung** [comment]
> imho mmap here should return the same provenance as the addr argument even though that addr argument wasn't valid until the mmap call.

That sounds pretty reasonable to me.
In particular you also must stay clear of any Rust-managed memory when doing this.

**Comment by algesten** [comment]
In the embedded code I've encountered (STM32 mainly), this is the [common pattern](https://github.com/stm32-rs/stm32f4xx-hal/blob/master/src/gpio.rs#L473).

```rust
unsafe {
    // NOTE(unsafe) this reference will only be used for atomic writes with no side effects.
    let rcc = &(*RCC::ptr());

    // Enable clock.
    $GPIOX::enable(rcc);
    $GPIOX::reset(rcc);
}
```

There will be an underlying crate that is mostly generated from the MCU vendor's SVD-files (often with some patch set with correction on top). The  maps each MCU peripheral to a specific memory address [`RCC::ptr()` etc](https://docs.rs/stm32f4/0.14.0/src/stm32f4/stm32f413/mod.rs.html#1395).

```rust
impl GPIOA {
    pub const PTR: *const gpioa::RegisterBlock = 0x4002_0000 as *const _;
    pub const fn ptr() -> *const gpioa::RegisterBlock {
        Self::PTR
    }
}
```

The pointer will be into some block, probably "Peripherals" below starting `0x4000_0000` (from [reference manual](https://www.st.com/resource/en/reference_manual/rm0454-stm32g0x0-advanced-armbased-32bit-mcus-stmicroelectronics.pdf)).

<img width="708" alt="image" src="https://user-images.githubusercontent.com/227204/176133141-8160f3e9-262a-4053-afe2-e5824e2434e9.png">

For the proposed provenance API, are we saying that we will declare the entire Peripheral block on startup and then take additional offsets from inside that block for the individual pointers?

Given the current embedded landscape, I believe that having an API to declare provenance for an individual pointer would be less of an upheaval than going via a block.

(As an aside, I'm also not keen on the name `alloc` for this – not sure what I'd call it though)

**Comment by Darksonn** [comment]
> I would probably go far enough to say that converting these pointers into a reference would never be correct, although I hesitate to call it completely undefined behaviour.

As long as the address doesn't require volatile to access it, I don't see any problem with creating references to them. In fact, I think `mmap` is a great example of that.

**Comment by LawrenceEsswood** [comment]
> A size/span restriction does not seem useful.

Can we revisit this? The CHERI crowd (and I'm sure others) want to attach sizes to all pointers. Specifying a size, with UB if accesses are made outside of it, would make the CHERI transition easier. On a CHERI machine, the rust runtime may not even have the capacity to construct a pointer valid for "all memory that is outside of what the Abstract Machine knows about".

**Comment by adamgreig** [comment]
@algesten 

> For the proposed provenance API, are we saying that we will declare the entire Peripheral block on startup and then take additional offsets from inside that block for the individual pointers?
> 
> Given the current embedded landscape, I believe that having an API to declare provenance for an individual pointer would be less of an upheaval than going via a block.

The current svd2rust-generated API has to change regardless*, so I wouldn't look too closely at how it works currently. The proposed API here just creates a pointer-with-provenance from an integer, so whether that's a pointer to a struct or a pointer to a single register or even a pointer to the entire peripheral memory region is somewhat immaterial to the eventual API in PACs, I think.

*: Currently it creates references to memory-mapped structs, but since LLVM is always allowed to dereference references whenever it likes, and for MMIO peripherals reading them may have side effects (and so all operations must be volatile), this isn't OK - svd2rust needs to swap to only use direct volatile access on raw pointers (possibly generated with an offset from the peripheral base) for each field.

**Comment by RalfJung** [comment]
> For the proposed provenance API, are we saying that we will declare the entire Peripheral block on startup and then take additional offsets from inside that block for the individual pointers?

I am not entirely sure what you mean.  Basically what we are talking about is replacing `0x4002_0000 as *const _` by `ptr::from_foreign_addr(0x4002_0000)`.

> Can we revisit this? The CHERI crowd (and I'm sure others) want to attach sizes to all pointers. Specifying a size, with UB if accesses are made outside of it, would make the CHERI transition easier. On a CHERI machine, the rust runtime may not even have the capacity to construct a pointer valid for "all memory that is outside of what the Abstract Machine knows about".

I think most of the code we are talking about here is extremely platform-specific and has no chance of being portably even across "normal" architectures, let alone CHERI.
Yes, CHERI also wants something that *does* take a size, but that sounds like a CHERI-specific API to me -- note that it also needs a "parent capability" or so that it can restrict to this range, which is not something that even exists in other architectures.

**Comment by Lokathor** [comment]
Having a way to specify a size is reasonable if the platform requires it. Most platforms don't require a size and on those platforms and so they shouldn't need to provide a size. We can and should probably "easily" cfg out the no-size methods from being available on CHERI and similar.

I don't think that llvm should be using the size info to determine if UB is happening or anything of that nature.

**Comment by LawrenceEsswood** [comment]
> Having a way to specify a size is reasonable if the platform requires it. Most platforms don't require a size and on those platforms and so they shouldn't need to provide a size. We can and should probably "easily" cfg out the no-size methods from being available on CHERI and similar.
> 
> I don't think that llvm should be using the size info to determine if UB is happening or anything of that nature.

But do they really not require it? These platforms very much **do** have sizes for their memory mapped registers. This information is just _possibly_ not in use by some compilers. Why shouldn't LLVM be free to use the information for safety / bug checking?

Other allocations nominally have a size, why make these "pre-existing" allocations special? If a particular compiler implementation wants to drop this information on the floor, then so be it, but I think it is meaningful to most back-ends. We should not be discarding useful information early, just because we expect some compilers won't end up using it. The fact we live in a world where the convention is to discard bounds information for speed should not dissuade us from designing interfaces as we intend them. Programmers intend to only use a pointer constructed from an integer for a fixed range of addresses, and so should state so, even if we don't make the checks for them.

And what really is the cost the other way? At the time I copy base addresses from a PDF, I can copy the length at the same time. Not very onerous. And now this code is more portable, open to any future compiler changes, and can be the target of future instrumentation.

**Comment by Lokathor** [comment]
What makes them special is that the sizes are often enough just imaginary, and as long as you write particular bits into each particular byte it often doesn't matter if you do it with four `strb` or two `strh` or one `str`. And so sometimes you might view the memory one way or another logically but that doesn't change the physical limits. And at this level of programming the developer should be bound only by the physical limits of their device.

And as Ralf slyly hinted at: adding UB makes *more* danger, not less.

**Comment by LawrenceEsswood** [comment]
I agree in principle UB is dangerous, but you are not exactly specifying an alternative defined behaviour. Rather, nothing is being specified at all with regards to the extent to which such a pointer can be used. Are you not already imagining making it UB to use such a pointer to access memory that is a part of the memory rust understands for its heap and stack? Surely if UB is so unsafe, we should allow that too?

Saying something is UB retains power for the compiler to make it trapping behaviour (which is safer), rather than having to allow such an access because that part of the specification was skipped, and some ill-intentioned programmer wrote a program that relied upon the behaviour.

**Comment by LawrenceEsswood** [comment]
But why not "as long as I stick to just the memory regions I intended to access when I made this cast"? Just a slight narrowing of what the pointer should be used for. Certainly not _less_ safe.

I think "whatever I want that would be legal in asm" is a terrible target for a language that purports to be fixing memory safety issues.

**Comment by Lokathor** [comment]
I suppose it depends on what you mean by allocated object.

**Comment by RalfJung** [comment]
> I agree in principle UB is dangerous, but you are not exactly specifying an alternative defined behaviour. Rather, nothing is being specified at all with regards to the extent to which such a pointer can be used. Are you not already imagining making it UB to use such a pointer to access memory that is a part of the memory rust understands for its heap and stack?

Yes. That is the amount of UB we need. Beyond that, it is also UB if the load causes a fault or data race. But everything else we can should not make UB.

We cannot really assign a lot of semantics to those accesses since that memory doesn't exist in the AM.

> as long as i stick to external memory regions, let me write rust code that does whatever thing i want that would be legal in asm code.

No, I don't think we want that. The resulting raw ptr looks like any other raw ptr, so the compiler will still make a bunch of assumptions about it -- no data races, for example.

Having a notion of memory access that can be optimized less (and possibly a pointer type that always uses such accesses) is a separate question.

> Should the semantics not be something along the lines of: when this machine started, there were a few preexisting, statically located and sized objects. This interface allows you to get a pointer to one of them, with the provenance of exactly one of those existing allocations. They behave the same rules as you normally expect, in that using a pointer with provenance from one to access another is UB.

I don't see a lot of benefit from this kind of UB.
We could say that "a pointer with provenance from one, when used to access another, is either fine or the compiler is allowed to catch this as a bug" (similar to how we specify arithmetic overflow).

But I am worried about even starting to have non-trivial notions of provenance for this. We are talking about very close-to-the-metal code here. I would like us to stay as close as we can to the metal in our spec. IMO we should only have provenance if the metal has it. Which, for almost all cases, means no provenance -- except for CHERI, which needs different APIs.

Concretely: if we have one of these preeexisting objects at 0x100..0x200, and someone does `from_foreign_addr(0x180)`, I think they should be allowed to do `offset(-0x10)` and access the memory at 0x170. I think they would be really surprised if we said this is UB, or the compiler is allowed to trap this as a bug. Hence I don't think we can meaningfully have bounds on these pointers.

Maybe someone wants to build a safer API on top of this that optionally tracks bounds or whatever. But I think that is a library thing, we don't want or need this on the lowest-level primitive.

Sorry, but CHERI is just too different from everything else to capture it in a common API here.

**Comment by Lokathor** [comment]
But Ralf, a data race is a property of an *access*, and we're not talking about accesses ;3

**Comment by RalfJung** [comment]
You were talking about accesses by saying "whatever thing i want that would be legal in asm code", weren't you?  So yes indeed that's off-topic in this thread. ;) https://github.com/rust-lang/unsafe-code-guidelines/issues/321 is a great thread for exactly that goal: "do whatever asm would do" (for accesses).

This thread is "do whatever asm would do" for int2ptr casts. ;)

**Comment by Lokathor** [comment]
No, I did mean like you said in the longer post, that you should be able to offset from wherever to wherever else and it's still a valid pointer.

**Comment by Lokathor** [comment]
Unless you're going to literally prevent people from using `usize as *mut T`, which you cannot do in rust 1.0, then you need to decide what function call the old style `as` expression expands to.This means that if size is used in general then the function call needs to infer the size (eg: size_of T or something). But then you're back with all the badness previously discussed.

**Comment by RalfJung** [comment]
> Unless you're going to literally prevent people from using usize as *mut T, which you cannot do in rust 1.0, then you need to decide what function call the old style as expression expands to.

Oh that is decided, at least for me. ;)  It's [`from_exposed_addr`](https://doc.rust-lang.org/nightly/std/ptr/fn.from_exposed_addr.html).

That's a big hammer though. This thread is about providing a less big hammer targeted for this specific use-case.

**Comment by Lokathor** [comment]
That's fine, but I'll note that `from_exposed_addr` doesn't take a span parameter.

**Comment by Dirbaio** [comment]
How would this interact with `expose_addr`/`from_exposed_addr`?

It probably shouldn't be UB to pass a foreign pointer to `expose_addr`. Otherwise foreign pointers would be too dangerous, it'd be too easy to pass them to code written with only normal pointers in mind.

Does this means then the "all foreign memory" allocation can itself be exposed or not, just like normal allocations? And when it has been exposed then it's OK to use `from_exposed_addr` on it?

Would it make sense to make it "exposed by default"? This would mean existing embedded code that does stuff like `0x4000_0000 as *const RegisterBlock` would keep being correct in Strict Provenance (assuming int2ptr `as` expands to `from_exposed_addr`). Perhaps it's a bit weird though, and it is inconsistent with how exposing "normal" allocations works.

**Comment by digama0** [comment]
@Dirbaio External allocations are always exposed, so you can use `from_exposed_addr` on them without issue. It doesn't even make sense to have an external allocation which is not exposed since then it wouldn't be external.

**Comment by RalfJung** [comment]
> Would it make sense to make it "exposed by default"? This would mean existing embedded code that does stuff like 0x4000_0000 as *const RegisterBlock would keep being correct in Strict Provenance (assuming int2ptr as expands to from_exposed_addr).

Yes that was my intent. We need to keep that code working for backwards-compatibility anyway.
Any "globally well-known provenance" should be considered exposed. The reason we need to distinguish "exposed" vs "non-exposed" is that when some code generates a fresh provenance in the privacy of its own function (e.g. for their stack variables, or via `malloc`), we need to make sure nobody else can just get hold of that provenance via `from_exposed_addr`.

**Comment by Dirbaio** [comment]
Makes sense.

Why do we need a `from_foreign_addr` then if `from_exposed_addr` already works, then? 

It could enable better optimizations, since the compiler knows `from_foreign_addr` will never touch exposed program memory. Would these make a difference in practice though? Most code out there will conform to Strict Provenance, so most program memory won't be exposed

**Comment by RalfJung** [comment]
It was mostly meant as an explicit bit of documentation, i.e., the programmer making their intent more clear.

**Comment by digama0** [comment]
Especially if it is "the programmer making their intent more clear" I think there is use for the size field. Since as far as AM operations are concerned, the intrinsic is already superfluous since `from_exposed_addr` also works. Unless you want to say that `from_foreign_addr` on an exposed Rust allocation is UB?

**Comment by A1-Triard** [comment]
@RalfJung , can you imagine a situation when difference between `from_exposed_addr`, and `from_foreign_addr` would mean something in practice? As longer I think about it, as more artificial `from_foreign_addr` seems to me. Maybe it is better to accept that such situation breaks strict provenance unavoidably.

**Comment by RalfJung** [comment]
It's not unavoidable though. :shrug: 

`from_exposed_addr` is a non-deterministic operation, i.e. executing it twice on the same input could produce two different results. So that could make a difference for the optimizer, when compared with `from_foreign_addr` which is deterministic.

**Comment by DemiMarie** [comment]
> > I agree in principle UB is dangerous, but you are not exactly specifying an alternative defined behaviour. Rather, nothing is being specified at all with regards to the extent to which such a pointer can be used. Are you not already imagining making it UB to use such a pointer to access memory that is a part of the memory rust understands for its heap and stack?
> 
> Yes. That is the amount of UB we need. Beyond that, it is also UB if the load causes a fault or data race. But everything else we can should not make UB.

Forbidding data races is okay.  Forbidding faults is in general _not_ okay, as it would prevent legitimate uses of faults such as pagable kernel heap memory.  That actually has an additional constraint, which is that it can be safe to access some memory memory at time A or at time C, but _not_ at time B (between A and C) because time B is one at which paging is not permitted.

In short, for kernels and similarly low-level code, one wants to be able to reorder ordinary non-volatile loads and stores across most operations, but _not_ across arbitrary function calls, because these might make the memory temporarily inaccessible.

**Comment by DemiMarie** [comment]
Another point to keep in mind: Accessing freed memory is UB.  However, the memory allocator (almost by definition) needs to do exactly that!

**Comment by RalfJung** [comment]
>  Forbidding faults is in general not okay,

Sorry, I should have been more precise -- by "fault" I meant something that the program can actually observe. If the kernel or some other entity does some magic in the background, that is fine. But, crucially, all memory accesses must succeed and not have any side-effect besides reading/writing the given memory location.

Otherwise, the compiler would not even be allowed to reorder reads.

> In short, for kernels and similarly low-level code, one wants to be able to reorder ordinary non-volatile loads and stores across most operations, but not across arbitrary function calls, because these might make the memory temporarily inaccessible.

They will have to add suitable barriers or other annotations then. Rust is allowed to reorder some reads around arbitrary function calls. (In fact even C is allowed to do that, if the read-from memory is private to the current function. Rust is just allowed to do it in more situations due to its aliasing restrictions.)

Anyway this is off-topic for this thread, since we are not talking about memory *accesses* here, only about how to get the *pointer* to some memory.

**Comment by scottmcm** [comment]
> Particularly, lots of existing and well working rust code doesn't bother with a "free" step at all when using hardware pointers.

I don't know that that means "alloc" is an inappropriate word, though.  We describe `static` as an "allocation" that's definitely never freed either.

**Comment by lukas-code** [comment]
Would this allow creating allocations at address zero?

[Currently, dereferencing a null pointer seems to be always UB, even if it was obtained with `from_exposed_addr(0)`.](https://rust.godbolt.org/z/z6oc78YEP)

**Comment by digama0** [comment]
I think this will always be UB. Creating the allocation is probably fine, but actually using `*ptr` where `ptr` is equal to `0` should be UB. If you want to dereference 0, you should use `asm!`.

**Comment by RalfJung** [comment]
> [Currently, dereferencing a null pointer seems to be always UB, even if it was obtained with from_exposed_addr(0).](https://rust.godbolt.org/z/z6oc78YEP)

Indeed, Rust assumes that there is never an allocation at 0, so there can't be an exposed provenance that is valid for that address, either.

**Comment by DemiMarie** [comment]
> I think this will always be UB. Creating the allocation is probably fine, but actually using `*ptr` where `ptr` is equal to `0` should be UB. If you want to dereference 0, you should use `asm!`.

Only non-`volatile` access should be UB.  If [my proposal for `volatile` semantics](https://github.com/rust-lang/unsafe-code-guidelines/issues/321) is accepted, one can use `volatile` to access any address whatsoever, including address 0 and even freed memory.  The reason this does not harm optimizations is that the return value of `ptr::read_volatile` in these situations is unspecified (_not_ undefined — it won’t be `undef` or `poison`).  Having to use `asm` for this would be silly.

**Comment by RalfJung** [comment]
The general conclusion in this thread regarding the original question seems to be: why have a `ptr::extern` or so if one can use `ptr::from_exposed_addr` instead? Here is one argument: stabilizing `from_exposed_addr` is further off than stabilizing the other functions, because of their rather unconventional semantics (angelic non-determinism) that should probably be studied a bit more before it gets enshrined in stable Rust. We are hence probably going to stabilize the core strict provenance functions (`addr`, `with_addr`, `map_addr`, `invalid`) before the once involving 'expose'.

Therefore, having a dedicated function for this case of accessing an 'external' address could help users of stable Rust away from `as` casts earlier. It could also help document intent, clearly saying that this is *not* relying on any address having been previously exposed, but instead relies on the address being 'external' to the Rust AM.

**Comment by Lokathor** [comment]
Currently, when using `as`, the user doesn't have to worry about if the usize is previously a runtime exposed pointer or a hardware address, they just do the `as` cast. This has *less* chance of screwing up, because it's one less thing that matters. So personally I wouldn't want to use `extern_addr` or anything like that, I'd just wait longer for `from_exposed_addr` to stabilize.

**Comment by LunarLambda** [comment]
Personally I think "each calls returns a unique provenance of a given size" makes the most sense. It covers the most use cases with the least API surface area. Whole memory regions, splitting memory regions, "magical" allocators, etc.

I'm not sure how useful it is to have "The One Unknowable Everything-Else Provenance" (i.e. no size parameter, single provenance for all "external"/non-rust addresses). Feels like that's only a very *slightly* smaller hammer than `from_exposed_addr`, and the only use it serves is "being able to arbitrarily offset into all possible external memory", which I think is basically never done outside of `asm`, which should already be covered by the `FFI-ish` provenance?

On that note, what about memory regions that can be relocated at runtime? Is this something that can be at all expressed under strict provenance (other than for discarding and re-creating the existing provenances)? I thought of `with_addr`, but the documentation says it's equivalent to `wrapping_offset`, which in turn says the pointer must not leave the memory range of the allocation, which I think implicitly assumes that the allocation cannot move?

**Comment by RalfJung** [comment]
Allocated objects in Rust cannot move. So such relocations (in general) have to be modeled like `realloc`, and they generate a fresh provenance, where all old pointers now have a dangling provenance that must not be used any more.

**Comment by joboet** [comment]
Supposing that this function did take a size, what would the semantics of a pointer created with size zero be? Would calling `offset(n)` with $n \ne 0$ on it be undefined behaviour, or would it behave the same as a pointer returned by `ptr::invalid` (or whatever that function will be named in the future)?

**Comment by LunarLambda** [comment]
Doesn't Rust disallow memory allocations of size 0 (which is why Vec/Box/etc have special cases for ZSTs)?

I would assume the same to apply here

**Comment by joboet** [comment]
[It does not](https://doc.rust-lang.org/nightly/std/ptr/index.html#safety).

**Comment by LunarLambda** [comment]
https://doc.rust-lang.org/nightly/std/alloc/trait.GlobalAlloc.html#safety-1 says it's UB to request allocations of size 0, which is different from creating an invalid pointer which is okay to be used for accesses of size 0

So my assumption is "Rust allocations cannot be zero-sized" would hold here?

**Comment by RalfJung** [comment]
*Heap* allocations of size zero cannot be created with our native allocator. However, stack variables and global `static` with size 0 can exist and those are allocations, too.

**Comment by programmerjake** [comment]
heap allocations of size zero can be created with some implementations of `malloc`, e.g.glibc 2.38:
https://sourceware.org/git/?p=glibc.git;a=blob;f=malloc/malloc.c;h=e2f1a615a4fc7b036e188a28de9cfb132b2351df;hb=36f2487f13e3540be9ee0fb51876b1da72176d3f#l115

**Comment by Lokathor** [comment]
If only some native allocators support it then rust overall must assume it's *not* supported.

**Comment by programmerjake** [comment]
> If only some native allocators support it then rust overall must assume it's _not_ supported.

yes, except that Rust can't declare that no zero-sized heap allocations are possible, they just can't be done with Rust's standard methods of allocating memory.

**Comment by RalfJung** [comment]
ptr::invalid is a reasonable implementation for a zero-sized allocation (that cannot be freed).

**Comment by RalfJung** [comment]
Okay, after we spent most of the discussion above clarifying various misunderstandings -- how do we stand wrt adding an explicit operation here? Yes, we could tell people to use `with_exposed_provenance` for MMIO addresses, but once https://github.com/rust-lang/rust/pull/130350 lands, `with_exposed_provenance` will have extremely vague docs that guarantee very little. In contrast, I think we could provide a `from_extern_addr` function that is way more definite in what it guarantees. Something like:

```rust
/// Converts an address to a pointer. The address must be referring to "external" memory, i.e. memory that is not
/// already part of a Rust allocated object.
///
/// Pointers derived from the result of this function may not be used to access any memory that Rust "knows about"
/// (such as `static`s including `extern static`s, stack variables, or heap allocations).
///
/// You may derive references
/// from the result of this function, but note that these references come with the usual Rust aliasing requirements:
/// mutable references must be unique, and shared references must be read-only (except for bytes inside `UnsafeCell`).
/// This means that when multiple calls to `from_extern_addr_mut` are used to obtain pointers to the same memory,
/// and one of these pointers is used to create a reference, it is Undefined Behavior to use any of the other pointers
/// in a way that conflicts with the requirements of this reference.
/// Furthermore, note that Rust references allow spurious memory accesses, so it is a bad idea to create references
/// to MMIO memory regions.
///
/// # Example
///
/// Using `from_extern_addr_mut` to access an MMIO register that is known to reside at a particular
/// offset within an MMIO block at a fixed address:
///
/// ```no_run
/// const MMIO_ADDR: usize = 0xaffe_0000;
/// const REGISTER_OFFSET: usize = 0xfe;
///
/// let ptr: *mut u8 = std::ptr::from_extern_addr_mut(MMIO_ADDR);
/// ptr.add(REGISTER_OFFSET).write(0x01);
/// ```
fn from_extern_addr_mut<T>(addr: usize) -> *mut T { ... }
```

I don't think we want this function to take a size, unless we are willing to say "any access outside that size is UB" -- which would make it harder for people to migrate from `as` casts to this method. The size will be defined implicitly by the pointers derived from the result of this function.

@rust-lang/opsem what do you think?

**Comment by DemiMarie** [comment]
@RalfJung What if the allocation _will_ back the stack, heap, or other memory locations Rust _does_ know about?  Normal code will never need to do that, but operating system code (such as kernels and threading libraries) absolutely will.

**Comment by Lokathor** [comment]
For first boot procedures on baremetal systems the answer is "you must do this in assembly, before entering Rust code". I suspect setting up a thread's stack and the global heap might have similar "don't let rust actually know you're doing it" rules.

**Comment by RalfJung** [comment]
@DemiMarie that memory needs to go through some sort of "phase transition", similar to memory returned by the global allocator, where it is removed from the "external memory allocation" it was in and instead becomes part of a newly born native Rust allocated object.

To me, the main thing I was unsure about in these docs is whether we want to describe this function as "create a new allocated object from outside-Rust memory", or "give access to some outside-Rust memory". The former would imply that you can only call this function once for each memory range, which seems too limiting. But at some point a Rust allocated object is logically being created, and I feel it would be good to talk about that -- I am just not sure how, since that action is entirely axiomatic, i.e. we never actually tell the compiler about it, we just "imagine" that such objects exist and then they appear (assuming there is actually accessible memory there and they are disjoint from stack/heap/statics).

**Comment by Lokathor** [comment]
Well if (1) this really is for MMIO and so (2) the target is only accessed with volatile, then do we need to consider it a standard rust allocated object at all?

Within the llvm lang ref it's stated that a volatile access of an address isn't used as evidence that a normal access of the address is also allowed. Do we want to try and accept a logic of this sort into rust?

**Comment by RalfJung** [comment]
I don't think it's just MMIO. It's also relevant for cases where particular things are stored at particular addresses, e.g. the interrupt vector or other kinds of communication with the firmware in a freestanding environment. Not sure if some OSes don't also put particular information (such as the environment or args block) at particular addresses.

**Comment by ketsuban** [comment]
I feel like magic addresses for communicating with firmware are spiritually MMIO, even if they're actually stored in normal memory. The edge case is probably something like the memory address `0x3007FFC` on the Game Boy Advance, which is _kind of_ like an interrupt vector but not really - the firmware is sitting on the actual hardware interrupt vector, and `0x3007FFC` is just an address in work RAM that the firmware happens to pick up and use to call user code when an interrupt fires.

**Comment by RalfJung** [comment]
> I feel like magic addresses for communicating with firmware are spiritually MMIO, 

One doesn't use `volatile` to talk to them, does one?

I would prefer to use MMIO only for memory that needs `volatile`. So no, I don't think we should call this "spiritually MMIO", that just causes confusion.

**Comment by bjorn3** [comment]
Addresses that are used for communication with the firmware may need volatile if the firmware accesses them from asynchronous smm interrupts, right?

**Comment by ketsuban** [comment]
> One doesn't use `volatile` to talk to them, does one?

I mean, I do. Rust eliding a subsequent access is *probably* fine when it's `0x3007FFC` where normal usage is setting it once and then forgetting about it; less so when I'm trying to signal that I handled an interrupt by setting a bit in the value at `0x3007FF8`.

**Comment by RalfJung** [comment]
> Addresses that are used for communication with the firmware may need volatile if the firmware accesses them from asynchronous smm interrupts, right?

That sounds more like a usecase for atomics and atomic_signal_fence / compiler_fence to me?

**Comment by xmh0511** [comment]
There also has a corner case, what if MMIO has a fixed address `0`? According to https://doc.rust-lang.org/std/ptr/fn.null.html
> This function is equivalent to zero-initializing the pointer: MaybeUninit::<*const T>::zeroed().assume_init(). The resulting pointer has the address 0.

> Accessing (loading from or storing to) a place that is dangling or based on a misaligned pointer.

This still falls into UB domain

**Comment by programmerjake** [comment]
> There also has a corner case, what if MMIO has a fixed address `0`?

you'd probably have to rely on inline assembly for that then.

**Comment by Lokathor** [comment]
Yeah, asm doesn't have a specific restriction against using 0 as an address like Rust does. Bonus: you know that it's outside of Rust's address space and can't interfere with Rust.

Aside: extremely silly that they zero initialize the pointer that way in the docs.

**Comment by RalfJung** [comment]
Accessing address `0` is a separate issue, see https://github.com/rust-lang/unsafe-code-guidelines/issues/29. That's off-topic here since it has a bunch of extra complications.
